### PR TITLE
$not fix.

### PR DESCRIPTION
**Pull request by calvintwr**

### CODE REVIEW

**Comment by calvintwr** [comment]
@mickhansen I believe the failing tests are not related to my changes.

Please review and merge.

File: lib/dialects/abstract/query-generator.js [hunk 1]
```diff
@@ -1702,15 +1702,21 @@ module.exports = (function() {
         })();
       }
 
-      // Do [] to $in/$notIn normalization
+      //Do [] to $in/$notIn normalization
       if (value && (!fieldType || !(fieldType instanceof DataTypes.ARRAY))) {
```
Comment (mickhansen) [inline lib/dialects/abstract/query-generator.js]: Lets keep this block as $in/$notIn normalization and do $not to $ne normalization in another block.

**Comment by mickhansen** [comment]
I've left a single comment. Aftewards please rebase and squash your commits to a single descriptive commit (a lot of dirty/pointless commits in this PR currently).

**Comment by calvintwr** [comment]
@mickhansen what is the reason for checking `value.$not` in `else if (value.$not && Array.isArray(value.$not))`?

`Array.isArray(value.$not)` will return false if `value.$not` regardless of whether `value.$not` is `undefined`, `false` or `true`. Hence I don't see why it needs to be there. I will remove it, but just in case it is something that I missed out, let me know I will put it back.

**Comment by mickhansen** [comment]
@calvintwr If fieldType is not a postgres array and the argument is $not and it's an array it needs to be transferred to $notIn.

File: lib/dialects/abstract/query-generator.js [hunk 2]
```diff
@@ -1702,18 +1702,24 @@ module.exports = (function() {
...
             $in: value
           };
-        } else if (value.$not && Array.isArray(value.$not)) {
+        } else if (Array.isArray(value.$not)) {
           value.$notIn = value.$not;
           delete value.$not;
         }
       }
 
+      //$not handling. Pass arrays $notIn. Pass non-null, non-bool to $ne.
```
Comment (mickhansen) [inline lib/dialects/abstract/query-generator.js]: `// normalize $not: non-bool|non-null to $ne

**Comment by mickhansen** [comment]
What i mean is that depending on the context people might use `$not: []` to mean notIn or be a NOT ARRAY operator in PG.

**Comment by mickhansen** [comment]
Might be a contrived case, but we have to support `$not: {$any: []}` aswell.
Btw all builds are failing, are you running the unit test suite?

**Comment by calvintwr** [comment]
mick i'm referring to using `else if (value.$not && Array.isArray(value.$not))` vs 'else if ( Array.isArray(value.$not))`. But its okay I left it there. I think the intent is to short circuit.

I don't know how to use the test suite locally tbh. anyway there are some other stuff causing it to fail.

**Comment by mickhansen** [comment]
No it's currently failing on the unit tests, which can be run with a simple `make test-unit`.
Every single build is failing: https://travis-ci.org/sequelize/sequelize/builds/53322287 ` TypeError: Cannot read property '$not' of null`.

**Comment by mickhansen** [comment]
@calvintwr oh you're right, that if statement could be made shorter :) sorry i misunderstood you.

**Comment by mickhansen** [comment]
Looks like you just fixed with (the issue was checking `value.$not` without making sure value existed).

**Comment by calvintwr** [comment]
yes i missed out that after i took it out of the earlier if {} branch. But it will work. the test will still fail because of something else. I'm trying to squash my commits.... doesn't seem to work. its still there.

i did `git rebase -i HEAD~10`. I will figure my way around....

**Comment by calvintwr** [comment]
its okay I left the $value in the condition statement there. i'm just thinking that it might be faster without it, but no biggie, microseconds...

**Comment by calvintwr** [comment]
@mickhansen squashed commits. don't know why the test fails. I check the tests errors, `query-generator.js` isn't in the call stack....

**Comment by mickhansen** [comment]
@calvintwr mssql is being quite the little annoyance currently :)

**Pull request merged**